### PR TITLE
Port the backend of FractionalMaxPool3d from TH to ATen

### PR DESCRIPTION
**Pull request by chandlerzuo**
1. Port the FractionalMaxPool3d implementation from THNN/THCUNN to ATen.
2. Expose this function to Python module nn.

### CODE REVIEW

**Comment by ezyang** [comment]
This PR is going to be in a similar situation to #15531 so I am going to hold off on reviewing it until we resolve the issues there.

File: test/common_nn.py [hunk 1]
```diff
@@ -1664,6 +1664,19 @@ def multimarginloss_weights_no_reduce_test():
...
     ),
+    dict(
+        constructor=lambda: nn.FractionalMaxPool3d(
+            2, output_ratio=0.5, _random_samples=torch.DoubleTensor(2, 4, 3).uniform_()),
+        input_size=(2, 4, 5, 5, 5),
+        fullname='FractionalMaxPool3d_ratio',
+    ),
+    dict(
+        constructor=lambda: nn.FractionalMaxPool3d((2, 2, 2), output_size=(
+            4, 4, 4), _random_samples=torch.DoubleTensor(2, 4, 3).uniform_()),
```
Comment (ezyang) [inline test/common_nn.py]: First off, it's awesome that you added these tests, thanks a lot :) (I didn't even realize that this function wasn't exposed from Python when I originally bootcamped this task). What would be really good is to add some tests which test for asymmetric kernel size. Something like in https://github.com/pytorch/pytorch/pull/15823

File: aten/src/ATen/native/FractionalMaxPool3d.cpp [hunk 2]
```diff
@@ -0,0 +1,420 @@
...
+  IntList pool_size,
+  IntList output_size,
+  const Tensor& randomSamples) {
+
+  int outputT = output_size[0];
+  int outputW = output_size[1];
+  int outputH = output_size[2];
+  int poolSizeT = pool_size[0];
+  int poolSizeW = pool_size[1];
+  int poolSizeH = pool_size[2];
```
Comment (ezyang) [inline aten/src/ATen/native/FractionalMaxPool3d.cpp]: Ordering here will need fixing (a test that takes tuple pool size will help you observe the problem)

File: aten/src/ATen/native/FractionalMaxPool3d.cpp [hunk 3]
```diff
@@ -0,0 +1,420 @@
...
+  const Tensor& input,
+  const Tensor& gradOutput_,
+  Tensor& gradInput,
+  IntList output_size,
+  IntList pool_size,
+  const Tensor& indices) {
+
+  int outputT = output_size[0];
+  int outputW = output_size[1];
+  int outputH = output_size[2];
```
Comment (ezyang) [inline aten/src/ATen/native/FractionalMaxPool3d.cpp]: will need reordering (also below)

File: aten/src/ATen/native/cuda/FractionalMaxPool3d.cu [hunk 4]
```diff
@@ -0,0 +1,410 @@
...
+    // grid handles overflow per each plane
+    int outputPlaneSize = output_.size(2) *
+      output_.size(3) * output_.size(4);
+    dim3 grid(
+      (outputPlaneSize + 127) / 128,
+      input_.size(1),
+      input_.size(0));
+    dim3 block(outputPlaneSize > 128 ? 128 : outputPlaneSize);
+
+    AT_DISPATCH_FLOATING_TYPES_AND_HALF(
```
Comment (ezyang) [inline aten/src/ATen/native/cuda/FractionalMaxPool3d.cu]: It looks like we lost the specializations for pool size 2-7, although I am not sure we actually care.

File: aten/src/ATen/native/cuda/FractionalMaxPool3d.cu [hunk 5]
```diff
@@ -0,0 +1,410 @@
...
+  IntList pool_size,
+  IntList output_size,
+  const Tensor& indices) {
+    int dimh = 1;
+    int dimw = 2;
+    int dimt = 3;
+
+    int outputT = output_size[2];
+    int outputW = output_size[1];
+    int outputH = output_size[0];
```
Comment (ezyang) [inline aten/src/ATen/native/cuda/FractionalMaxPool3d.cu]: heh, this is an even weirder ordering to have picked! It should be T (0), H (1), W (2)

File: aten/src/ATen/native/cuda/FractionalMaxPool3d.cu [hunk 6]
```diff
@@ -0,0 +1,410 @@
...
+      "gradOutput width unexpected"
+    );
+    AT_CHECK(
+      outputT == gradOutput.size(dimt),
+      "fractional_max_pool3d_backward_out_cuda_template(): ",
+      "gradOutput time unexpected"
+    );
+
+    /* resize */
+    gradInput = at::zeros_like(input);
```
Comment (ezyang) [inline aten/src/ATen/native/cuda/FractionalMaxPool3d.cu]: you haven't fixed this one

File: aten/src/ATen/native/cuda/FractionalMaxPool3d.cu [hunk 7]
```diff
@@ -0,0 +1,410 @@
...
+
+    auto gradInput_ = gradInput;
+    auto gradOutput_ = gradOutput;
+    auto indices_ = indices;
+
+    if(ndims == 4) {
+      gradInput_.reshape({1, gradInput.size(0), inputH, inputW, inputT});
```
Comment (ezyang) [inline aten/src/ATen/native/cuda/FractionalMaxPool3d.cu]: and these need fixing too

File: aten/src/ATen/native/cuda/FractionalMaxPool3d.cu [hunk 8]
```diff
@@ -0,0 +1,410 @@
...
+      "fractional_max_pool3d_backward_out_frame",
+      [&] {
+        fractional_max_pool3d_backward_out_frame<scalar_t>
+        <<<grid, block, 0, at::cuda::getCurrentCUDAStream()>>>(
+          gradInput_.packed_accessor<scalar_t, 5>(),
+          gradOutput_.packed_accessor<scalar_t, 5>(),
+          indices_.packed_accessor<int64_t, 5>()
+        );
+      }
+    );
```
Comment (ezyang) [inline aten/src/ATen/native/cuda/FractionalMaxPool3d.cu]: Missing cudaGetLastError check

File: aten/src/ATen/native/cuda/FractionalMaxPool3d.cu [hunk 9]
```diff
@@ -0,0 +1,410 @@
...
+  }
+
+Tensor& fractional_max_pool3d_backward_out_cuda(
+  at::Tensor& gradInput,
+  const at::Tensor& gradOutput_,
+  const at::Tensor& input,
+  IntList pool_size,
+  IntList output_size,
+  const at::Tensor& indices) {
+    gradInput.resize_as_(input);
```
Comment (ezyang) [inline aten/src/ATen/native/cuda/FractionalMaxPool3d.cu]: unnecessary resize

File: torch/nn/functional.py [hunk 10]
```diff
@@ -368,6 +368,76 @@ def _fractional_max_pool2d(input, kernel_size, output_size=None,
...
+    Fractional MaxPooling is described in detail in the paper `Fractional MaxPooling`_ by Ben Graham
+
+    The max-pooling operation is applied in :math:`kH \times kW \times kT` regions by a stochastic
+    step size determined by the target output size.
+    The number of output features is equal to the number of input planes.
+
+    Args:
+        kernel_size: the size of the window to take a max over.
+                     Can be a single number :math:`k` (for a square kernel of :math:`k \times k \times k`)
+                     or a tuple :math:`(kH \times kW \times kT)`
```
Comment (ezyang) [inline torch/nn/functional.py]: Once we reorder the parameters, the docs here should change to give the conventional ordering TxHxW

**Comment by ezyang** [comment]
I finished a complete review; I think my comments are the only outstanding issues.

File: aten/src/ATen/native/FractionalMaxPool3d.cpp (hunk 2 above)
Comment (chandlerzuo) [reply aten/src/ATen/native/FractionalMaxPool3d.cpp]: Fixed

File: aten/src/ATen/native/FractionalMaxPool3d.cpp [hunk 11]
```diff
@@ -0,0 +1,420 @@
...
+
+template<typename scalar_t>
+static void fractional_max_pool3d_backward_out_single_batch_frame(
+  scalar_t* gradInput,
+  scalar_t* gradOutput,
+  int64_t* indices,
+  int numPlanes,
+  int inputT, int inputW, int inputH,
+  int outputT, int outputW, int outputH) {
+  int plane;
```
[the remark this answers is not in the document]
Comment (chandlerzuo) [reply aten/src/ATen/native/FractionalMaxPool3d.cpp]: I have changed all int back to int64_t. I think it is very unlikely to have these variables overflow for real data, but agree that we need to make this error prone.

File: aten/src/ATen/native/cuda/FractionalMaxPool3d.cu (hunk 4 above)
Comment (chandlerzuo) [reply aten/src/ATen/native/cuda/FractionalMaxPool3d.cu]: It is still preserved in the fractional_max_pool3d_out_frame function.

File: aten/src/ATen/native/cuda/FractionalMaxPool3d.cu (hunk 6 above)
Comment (chandlerzuo) [reply aten/src/ATen/native/cuda/FractionalMaxPool3d.cu]: Fixed.

File: aten/src/ATen/native/cuda/FractionalMaxPool3d.cu (hunk 8 above)
Comment (chandlerzuo) [reply aten/src/ATen/native/cuda/FractionalMaxPool3d.cu]: Fixed.

File: aten/src/ATen/native/cuda/FractionalMaxPool3d.cu (hunk 7 above)
Comment (chandlerzuo) [reply aten/src/ATen/native/cuda/FractionalMaxPool3d.cu]: Fixed.

File: aten/src/ATen/native/cuda/FractionalMaxPool3d.cu (hunk 9 above)
Comment (chandlerzuo) [reply aten/src/ATen/native/cuda/FractionalMaxPool3d.cu]: Fixed.

File: test/common_nn.py (hunk 1 above)
Comment (chandlerzuo) [reply test/common_nn.py]: The asymmetric test is added as suggested.

**Review by ezyang — APPROVED** [review]
Make sure CI passes before landing

**Pull request closed without merging**